File: 0063-unique-paths-ii/0063-unique-paths-ii.cpp
class Solution {
    int solve(vector<vector<int>>& Grid,int row,int col){
        vector<vector<int>>dp(row,vector<int>(col,-1));

        
        for(int i=0;i<row;i++){
            for(int j=0;j<col;j++){
                if(i==0 && j==0){
                    dp[i][j]=1;
                }
                else{
                
                int up=0;
                if(i>0 && Grid[i-1][j]!=1){
                    up=dp[i-1][j];
                }
                int left=0;
                 if(j>0 && Grid[i][j-1]!=1){
                    left=dp[i][j-1];
                    }
                dp[i][j]=left+up;
                
            }}
        }
        
        
        return dp[row-1][col-1];
    }
public:
    int uniquePathsWithObstacles(vector<vector<int>>& Grid) {
        int row=Grid.size();
        int col=Grid[0].size();
        if(Grid[row-1][col-1]==1) return 0;
        return solve(Grid,row,col);
    }
};

File: 0063-unique-paths-ii/0063-unique-paths-ii.java
class Solution {
    
    public int f(int[][] nums,int[][] dp,int i,int j){ 
        if(i==nums.length-1&&j==nums[0].length-1){ 
            if(nums[i][j]==0) return 1;
            return 0;
        }
        if(i==nums.length||j==nums[0].length) return 0;
        if(dp[i][j]!=-1) return dp[i][j];
        if(nums[i][j]==1){dp[i][j]=0; return 0;}
        int x=f(nums,dp,i+1,j)+f(nums,dp,i,j+1);
        dp[i][j]=x;
        return dp[i][j];
    }
    public int uniquePathsWithObstacles(int[][] nums) {
        int[][] dp=new int[nums.length][nums[0].length];
        for(int[] e:dp) Arrays.fill(e,-1);
        return f(nums,dp,0,0);
    }
}